src/scanner.c

```diff
@@ -0,0 +1,640 @@
+
+            while (length < info->delimiter_length) {
+                advance(lexer);
+                if (lexer->lookahead == info->end_delimiter) {
+                    length++;
+                } else {
+                    break;
+                }
+            }
+
+            if (length == info->delimiter_length && (!is_heredoc || newline)) {
+                return has_content;
+            }
+        } else {
+            if (lexer->lookahead == '#') {
+                advance(lexer);
+                if (info->supports_interpol && lexer->lookahead == '{') {
+                    return has_content;
+                }
+            } else if (lexer->lookahead == '\\') {
+                advance(lexer);
+                if (is_heredoc && lexer->lookahead == '\n') {
+                    // We need to know about the newline to correctly recognise
+                    // heredoc end delimiter, so we intentionally ignore
+                    // escaping
+                } else if (info->supports_interpol ||
+                           lexer->lookahead == info->end_delimiter) {
+                    return has_content;
+                }
+            } else if (lexer->lookahead == '\0') {
+                // If we reached the end of the file, this means there is no
+                // end delimiter, so the syntax is invalid. In that case we
+                // want to treat all the scanned content as quoted content.
+                return has_content;
+            } else {
+                advance(lexer);
+            }
+        }
+    }
+
+    return false;
+}
+
+static bool scan_newline(TSLexer *lexer, const bool *valid_symbols) {
+    advance(lexer);
+
+    while (is_whitespace(lexer->lookahead)) {
+        advance(lexer);
+    }
+
+    // Note we include all the whitespace after newline, so that the
+    // parser doesn't have to go through it again
+    lexer->mark_end(lexer);
+
+    if (lexer->lookahead == '#') {
+        lexer->result_symbol = NEWLINE_BEFORE_COMMENT;
+        return true;
+    }
+
+    if (lexer->lookahead == 'd' && valid_symbols[NEWLINE_BEFORE_DO]) {
+        lexer->result_symbol = NEWLINE_BEFORE_DO;
+        advance(lexer);
+        if (lexer->lookahead == 'o') {
+            advance(lexer);
+            return is_token_end(lexer->lookahead);
+        }
+        return false;
+    }
+
+    if (valid_symbols[NEWLINE_BEFORE_BINARY_OPERATOR]) {
+        lexer->result_symbol = NEWLINE_BEFORE_BINARY_OPERATOR;
+
+        // &&, &&&
+        if (lexer->lookahead == '&') {
+            advance(lexer);
+            if (lexer->lookahead == '&') {
+                advance(lexer);
+                if (lexer->lookahead == '&') {
+                    advance(lexer);
+                    return check_operator_end(lexer);
+                } else {
+                    return check_operator_end(lexer);
+                }
+            }
+            // =, ==, ===, =~, =>
+        } else if (lexer->lookahead == '=') {
```

Comment:
Ideally, I would keep the original formatting :)

Reply:
it'd be nice if there was a clang-format file for that standard..having to manually do that is a bit annoying :sweat_smile: if the main want is indents being 2 spaces then thats much easier to do

Reply:
To be fair there is no standard, I didn't use a formatter :D

If there's an easy way to add a formatter task then that works for me (otherwise whoever changes the file may reformat it using different tool or IDE). Perhaps the `clang-format` npm package would do? Then we could add it to the formatting tasks we have:

https://github.com/elixir-lang/tree-sitter-elixir/blob/4ba9dab6e2602960d95b2b625f3386c27e08084e/package.json#L19-L20

Reply:
auto formatting happens from the clangd lsp, a `.clang-format` file in the project root can set this up easily and clangd will pick it up and override the users config. Here's my config for example to show you a few options that can be configured:

```yaml
AlignArrayOfStructures: Left
BasedOnStyle: LLVM
IndentCaseLabels: true
IndentGotoLabels: true
IndentPPDirectives: AfterHash
IndentWidth: 4
KeepEmptyLinesAtTheStartOfBlocks: false
SeparateDefinitionBlocks: Always
SortIncludes: CaseInsensitive
SpaceAfterCStyleCast: false
SpaceAfterLogicalNot: false
SpaceBeforeCaseColon: false
```

Rest are here https://clang.llvm.org/docs/ClangFormatStyleOptions.html

Reply:
hmm..sure that could be fine - I'm not against it but a clang-format file could be enough if anyone's using the lsp/format

Reply:
Sure, but with the file alone we can't enforce it on CI, and we assume they use LSP (which for example I don't :)).

Reply:
Yep..understandable. Added the dep + format script, but it has no easy "just show me the errors" option so I just used diff (returns 1 on differences, 0 if none)